internal/pkg/peer/blocksprovider/block_receiver.go

```diff
@@ -0,0 +1,171 @@
+	"github.com/hyperledger/fabric-protos-go/gossip"
+	"github.com/hyperledger/fabric-protos-go/orderer"
+	"github.com/hyperledger/fabric/common/flogging"
+	gossipcommon "github.com/hyperledger/fabric/gossip/common"
+	"github.com/hyperledger/fabric/internal/pkg/peer/orderers"
+	"github.com/pkg/errors"
+)
+
+type BlockReceiver struct {
+	channelID           string
+	gossip              GossipServiceAdapter
+	blockGossipDisabled bool
+	blockVerifier       BlockVerifier
+	deliverClient       orderer.AtomicBroadcast_DeliverClient
+	cancelSendFunc      func()
+	recvC               chan *orderer.DeliverResponse
+	stopC               chan struct{}
+	endpoint            *orderers.Endpoint
+
+	mutex    sync.Mutex
+	stopFlag bool
+
+	logger *flogging.FabricLogger
+}
+
+// Start starts a goroutine that continuously receives blocks.
+func (br *BlockReceiver) Start() {
+	br.logger.Infof("Starting to receive")
+	go func() {
+		for {
+			resp, err := br.deliverClient.Recv()
+			if err != nil {
+				br.logger.Warningf("Encountered an error reading from deliver stream: %s", err)
+				close(br.recvC)
+				return
+			}
+
+			select {
+			case br.recvC <- resp:
+			case <-br.stopC: // local stop signal
+				close(br.recvC)
+				return
+			}
+
+		}
+	}()
+}
+
+func (br *BlockReceiver) Stop() {
+	if br == nil {
+		return
+	}
+
+	br.mutex.Lock()
+	defer br.mutex.Unlock()
+
+	if br.stopFlag {
+		return
+	}
+
+	br.stopFlag = true
+	close(br.stopC)
+}
+
+// ProcessIncoming processes incoming messages until stopped or encounters an error.
+func (br *BlockReceiver) ProcessIncoming(onSuccess func(blockNum uint64)) error {
+	var err error
+
+RecvLoop: // Loop until the endpoint is refreshed, or there is an error on the connection
+	for {
+		select {
+		case <-br.endpoint.Refreshed:
+			br.logger.Infof("Ordering endpoints have been refreshed, disconnecting from deliver to reconnect using updated endpoints")
+			err = &errRefreshEndpoint{message: fmt.Sprintf("orderer endpoint `%s` has been refreshed, ", br.endpoint.Address)}
+			break RecvLoop
+		case response, ok := <-br.recvC:
+			if !ok {
+				br.logger.Warningf("Orderer hung up without sending status")
+				err = errors.Errorf("orderer `%s` hung up without sending status", br.endpoint.Address)
+				break RecvLoop
+			}
+			var blockNum uint64
+			blockNum, err = br.processMsg(response)
+			if err != nil {
+				br.logger.Warningf("Got error while attempting to receive blocks: %v", err)
+				err = errors.WithMessagef(err, "got error while attempting to receive blocks from orderer `%s`", br.endpoint.Address)
+				break RecvLoop
+			}
+			onSuccess(blockNum)
+		case <-br.stopC:
+			br.logger.Infof("BlockReceiver got a signal to stop")
+			err = &errStopping{message: "got a signal to stop"}
+			break RecvLoop
+		}
+	}
+
+	// cancel the sending side and wait for the start goroutine to exit
+	br.cancelSendFunc()
+	<-br.recvC
+
+	return err
+}
+
+func (br *BlockReceiver) processMsg(msg *orderer.DeliverResponse) (uint64, error) {
```

Comment:
please let's not do it this way. 
We need a component we can re-use for orderer too. 

Here, we couple relaying the block to gossip with retrieving it from the orderer.

Reply:
Done